WalletWasabi/WabiSabi/Backend/Rounds/Arena.cs

```diff
@@ -366,22 +367,25 @@ public async Task RemoveInputAsync(InputsRemovalRequest request, CancellationTok
 
 		public async Task<ConnectionConfirmationResponse> ConfirmConnectionAsync(ConnectionConfirmationRequest request, CancellationToken cancellationToken)
 		{
+			Round? round;
+			Alice? alice;
+			var realAmountCredentialRequests = request.RealAmountCredentialRequests;
+			var realVsizeCredentialRequests = request.RealVsizeCredentialRequests;
+
 			using (await AsyncLock.LockAsync(cancellationToken).ConfigureAwait(false))
 			{
-				if (Rounds.FirstOrDefault(x => x.Id == request.RoundId) is not Round round)
+				round = Rounds.FirstOrDefault(x => x.Id == request.RoundId);
+				if (round is null)
 				{
 					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.RoundNotFound, $"Round ({request.RoundId}) not found.");
 				}
 
-				var alice = round.Alices.FirstOrDefault(x => x.Id == request.AliceId);
+				alice = round.Alices.FirstOrDefault(x => x.Id == request.AliceId);
 				if (alice is null)
 				{
 					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.AliceNotFound, $"Round ({request.RoundId}): Alice ({request.AliceId}) not found.");
 				}
 
```

Comment:
```suggestion

				if (alice.ConfirmedConnection)
				{
					Prison.Note(alice, round.Id);
					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.DuplicateConfirmation);
				}

```

If this is seen, it's clearly a malicious action because an idempotent request would be handled earlier. At this point this `alice` is already committed by being previously confirmed. We can either fail the round and ban it immediately, or we can let it stay, but we should expect it to disrupt the signing phase as well. If we remove it, the credentials were already issued, so this client could register illegal outputs.

See also https://github.com/zkSNACKs/WalletWasabi/pull/6154#discussion_r684533999.

Reply:
(note this is arguably out of scope and might be more appropriate in a separate PR, since it will also need tests, etc)

Reply:
Actually, it's not as bad as I thought, because `MultiPartyTransactionState` forbids duplicate inputs, so this will cause an error at line 446 instead of issuing duplicate credentials, even though two connection confirmation requests can race each other.

That said, this is really subtle, and I think is the strongest argument we have so far for making `Alice` objects drive their own state forward. If `Arena` ensures prevout uniqueness, and `Alice` is locked for each relevant request, it could ensure that the per-alice state machine is linear and easy to follow.

Also, I think that the fact that the design of MPTS prevented this DoS vector by being a standalone component that focuses *only* on ensuring the transaction is valid & standard demonstrates the advantage of separating concerns, and how `Arena` controlling everything is a liability.

Reply:
https://github.com/zkSNACKs/WalletWasabi/pull/6162